File: strategy-game-backend/src/main/kotlin/de/ruegnerlukas/strategygame/backend/external/api/message/handler/MessageHandler.kt
package de.ruegnerlukas.strategygame.backend.external.api.message.handler

import de.ruegnerlukas.strategygame.backend.external.api.message.models.Message
import de.ruegnerlukas.strategygame.backend.external.api.message.models.SubmitTurnMessage
import de.ruegnerlukas.strategygame.backend.ports.provided.turn.TurnSubmitAction
import de.ruegnerlukas.strategygame.backend.shared.Logging

/**
 * Message-dispatcher for websocket messages
 */
class MessageHandler(
    private val turnSubmitAction: TurnSubmitAction
) : Logging {

    /**
     * Called for any incoming message
     * @param message the message
     */
    suspend fun onMessage(message: Message<*>) {
        log().info("Received message '${message.type}' from connection ${message.meta?.connectionId}")
        when (message.type) {
            SubmitTurnMessage.TYPE -> handleSubmitTurn(message as SubmitTurnMessage)
            else -> log().info("Unknown message type: ${message.type}")
        }
    }

    private suspend fun handleSubmitTurn(message: SubmitTurnMessage) {
        turnSubmitAction.perform(message.meta!!.userId, message.meta!!.gameId, message.payload.commands.map { it.asCommandData() })
    }

}

File: strategy-game-frontend/src/external/api/messaging/messageHandler.ts
import {SetGameStateAction} from "../../../core/actions/gamelobby/setGameStateAction";
import {PayloadGameState} from "../../../models/messaging/payloadGameState";


export class MessageHandler {

    private readonly gameSetState: SetGameStateAction;

    constructor(gameSetState: SetGameStateAction) {
        this.gameSetState = gameSetState;
    }


    onMessage(type: string, payload: any): void {
        console.log("Received message", type, payload);
        if (type === "game-state") {
            this.onWorldState(payload);
        }
    }

    onWorldState(payload: PayloadGameState) {
        this.gameSetState.perform(payload.game);
    }

}